Signing of message as JWS structure with JWT Claims set as payload.

sbhatore95 opened the pull request:
Any message can be signed now in JWS format with expiration
enforced with it so that messages are expired on time, and
purpose so that a message doesn't get misused in place for
other message. Use of expiration and purpose are optional.
The payload for JWS is the JWT Claims set containing
'pld' - the message, 'exp' - expiration and 'for' - purpose/scope
of the message. JWT Claims is not fully implemented yet as
per the spec here:
http://tools.ietf.org/html/draft-ietf-oauth-json-web-token-13#section-4
This is just a basic implementation of JWT/JWS which could
be extended further.

sikachu commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+    # Keyword argument +:payload+ is mandatory
+    def initialize(payload:, **options)
+      @payload = payload
+      @purpose = self.class.pick_purpose(options)
```
Is there a reason why `pick_purpose` has to be a class method? I feel like you could just inline this and maybe add a constant for `DEFAULT_PURPOSE` or something.

``` ruby
@purpose = options[:for] || DEFAULT_PURPOSE
```

sikachu commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+      end
+
+      def verify!(claims, options = {})
+        raise InvalidClaims if claims[:for] != pick_purpose(options)
```
- You can inline this and use a constant instead.
- This method is a bit hard to understand since there're two tailing ifs. Consider making them as a proper `if/else` block will make the code cleaner.

sikachu commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+
+      private
+        def parse_expiration(expiration)
+          return true unless expiration
```
This method could be cleanup a bit as well to make it easier to read and follow.

sikachu commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +65,97 @@ def valid_message?(signed_message)
+      data = signed_message.split('.')[1]
+      @serializer.load(decode(data)).symbolize_keys
+    rescue ArgumentError => argument_error
+      return if argument_error.message =~ %r{invalid base64}
```
You can just do:

``` ruby
raise unless argument_error.message =~ %r{invalid base64}
```

sikachu commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+
+      def verify!(claims, options = {})
+        raise InvalidClaims if claims[:for] != pick_purpose(options)
+        claims[:pld] if parse_expiration(claims[:exp])
```
Question: Is `pld` something that we introduced, or is that something that is already there? I'm more incline to just make it `payload` and `expiration` if it's something we could change.

sbhatore95 replied:
`pld` is something we introduced. `exp` is actually there in the JWT spec so we can't go with `expiration`. All other parameter names are also three lettered. Here, http://self-issued.info/docs/draft-ietf-oauth-json-web-token.html. It would be odd to have `payload` there. I think we can go with `pld`. :smile:

sikachu commented on activesupport/test/claims_test.rb:
```diff
@@ -0,0 +1,154 @@
+
+  def test_equal_only_with_same_purpose
+    assert_equal @claims, ActiveSupport::Claims.new(payload: @value, for: 'test')
+    refute_equal @claims, ActiveSupport::Claims.new(payload: @value, for: 'login')
```
We [don't use refute](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#follow-the-coding-conventions) in the code base. You want to use `assert_not_equal` here.

sikachu commented on activesupport/test/message_verifier_test.rb:
```diff
@@ -16,7 +16,122 @@ def load(value)
+
+  def test_valid_message
+    header, claims, digest = @verifier.generate(@value, @options).split(".")
+    assert !@verifier.valid_message?(nil)
```
Use `assert_not` for these.

sbhatore commented:
@sikachu I have made a few changes after your review. I think we can now open a PR in rails/master. What do you say @kaspth ?

kaspth commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+        return options[:expires_at] if options.key?(:expires_at)
+
+        # Expiration defaults to 1 month through class level
+        # variable +:expires_in+ set in railtie.
```
This shouldn't be in a private method's definition. It would be better in the general Claims documentation.

kaspth commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+module ActiveSupport
+  # +Claims+ supports <tt>ActiveSupport::MessageVerifier</tt> by
+  # representation of message in the form JWT Claims and verifying
+  # the claims of the signed message. 
```
This documentation is extremely meager and doesn't really make sense to me. If you can't come up with anything better, let's cut it.

kaspth commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+    class ExpiredClaims < StandardError; end
+    
+    attr_reader :payload, :purpose, :expires_at
+    DEFAULT_PURPOSE = 'universal'
```
I'd rather we stay consistent and called it `UNIVERSAL_PURPOSE`.

kaspth commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,63 @@
+    attr_reader :payload, :purpose, :expires_at
+    DEFAULT_PURPOSE = 'universal'
+
+    # Keyword argument +:payload+ is mandatory
```
In general we assume people know their Ruby, and so we don't have to explain what a required keyword argument is :smile:

sbhatore95 replied:
:sweat_smile:  :stuck_out_tongue:

kaspth replied:
Because it'll raise if people don't pass a payload argument:

``` irb
irb(main):001:0> class Claims
irb(main):002:1>   def initialize(payload:, **options)
irb(main):003:2>   end
irb(main):004:1> end
=> :initialize
irb(main):005:0> Claims.new
ArgumentError: missing keyword: payload
    from (irb):2:in `initialize'
    from (irb):5:in `new'
    from (irb):5
    from /Users/kasperhansen/.rbenv/versions/2.2.2/bin/irb:11:in `<main>'
```

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -46,16 +46,17 @@ def initialize(secret, options = {})
-    #   verifier.verified(signed_message) # => 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_at: 1.day.from_now)
+    #   verifier.verified(signed_message)
+    #   # => {:pld=>"a private message", :for=>"test", :exp=>"2015-06-29T07:16:12.579Z"}
```
This breaks backwards compatibility. This should absolutely not change :smile:

I'm curious how you arrived at this as the acceptable return value? Perhaps there's something we haven't talked about enough :smile:

See, I view the JWT generation as an implementation detail that helps us support more options. My stance is that people shouldn't have to care about how we accomplish this.

sbhatore95 replied:
Hmmm...this is some new learning. I didn't know that each function's return value was to be taken into consideration. I will fix this. :smile:

kaspth replied:
I'm sorry you've been caught off guard by this. Weren't we clear enough about wanting to maintain backwards compatibility? Because wanting backwards compatibility automatically means we have to consider what methods return (and more).

As an aside, I want to talk more generally about how you respond to feedback. I feel like it's been happening a lot that you respond extremely confidently, but you almost don't ask any questions. Let's take this comment thread as an example. You've just seen something that you didn't know you had to be aware of and you have no questions about that at all? There's really no other things you think should be cleared up before you start working on this? I will be honest and say that worries me a bit :smile:

I guess my problem is that I almost never hear what you're **thinking**, you just do stuff. Then I tell you to change something and you change it. I would rather, we could learn to anticipate each other, so we didn't need so many cycles to get somewhere. Does that make sense? :smile:

sbhatore95 replied:
I am sorry I got didn't think much here :see_no_evil: . I think I know why we are trying to do backward compatibility. You can say that I focused more on getting the output from the whole system rather than thinking about each component and overall scenario.

I will be more careful in future. :purple_heart: :blue_heart: :green_heart: :heart: :sparkling_heart:

kaspth replied:
That's okay, you're learning :smile:

Just to clarify we want backwards compatibility for ease of upgrading. If there's no breaking changes a user has a smooth upgrade. That's especially important for something security related, which can sometimes feel like extra hassle for little gain (security isn't always the most fun problem to handle).

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +65,96 @@ def valid_message?(signed_message)
     end
 
-    # Decodes the signed message using the +MessageVerifier+'s secret.
+    # Decodes the signed message and legacy signed message using the +MessageVerifier+'s secret.
```
Whenever you're writing something, it's important to try to anticipate your reader. Here they're likely to ask: What's a legacy signed message?

Try reading through your documentation it's extremely hard to get an answer to that as a reader.

Also neither purpose or expiration is given a reason to exist. They're just shown. That's also tough to for a reader to understand. Ask yourself this: we're making these changes to improve security, so how can we let people how to use these options to improve their app's security?

kaspth replied:
I've said this a few times before, but please do take a look at how Global ID's signed version does this. The README explains the reasoning behind setting purpose and expiration: https://github.com/rails/globalid#signed-global-ids

sbhatore95 replied:
Nice point. I will add more details and make it easier for the reader to grasp this. :smile:

kaspth commented on activesupport/test/claims_test.rb:
```diff
@@ -0,0 +1,154 @@
+
+class ClaimsExpirationTest < ActiveSupport::TestCase
+  def setup
+    @value = 'payload'
```
I feel like you're flip flopping between value and payload. Let's stick to payload as much as we can :smile:

kaspth commented on activesupport/test/message_verifier_test.rb:
```diff
@@ -16,18 +16,132 @@ def load(value)
 
   def setup
-    @verifier = ActiveSupport::MessageVerifier.new("Hey, I'm a secret!")
+    @secret = 'Hey, I\'m a secret!'
```
Let's use a double quoted string here to avoid `\`.

sbhatore95 commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,156 @@ def valid_message?(signed_message)
+      @claims = Claims.new(payload: value, **options)
+      data = [encode(serialized_header), encode(serialized_claims)].join('.')
+      "#{data}.#{encode(generate_digest(to_ascii(data)))}"
     end
```
I missed this in the documentation. Input to signing data should be converted to ASCII characters.
Here http://tools.ietf.org/html/rfc7515#section-5.1 point 5.

kaspth replied:
Aren't base64 all ASCII characters? Other implementations don't seem to do it: https://github.com/progrium/ruby-jwt/blob/master/lib/jwt.rb

sbhatore95 replied:
Yeah. They are using some other conversions. I can't figure out why as of now.
I think we can go just with `::Base64.strict_encode64` function instead of the one in this https://github.com/progrium/ruby-jwt/blob/master/lib/jwt.rb.

kaspth replied:
Sounds good :+1:

sbhatore95 commented on activesupport/test/message_verifier_test.rb:
```diff
@@ -16,18 +16,133 @@ def load(value)
+    assert_equal @payload, @verifier.verify(message, for: 'test')
+  end
+
+  def raise_invalid_signature_when_value_is_nil
```
Here's a doubt:
- **Case 1: Previous `MessageVerifer`**
  Accepts `nil` as an argument in `generate` method  and returns a signed message. On calling `verify` on that signed_message, it raises `ActiveSupport::MessageVerifier::InvalidSignature` even when no tampering is done.
- **Case 2: New `MessageVerifier`** 
  I have implemented it same as previous `MessageVerifier` as shown in this test.

However, I don't think `generate` should accept `nil` as a value at all. It doesn't make sense to sign nothing. What do you think @kaspth ?

kaspth replied:
There was some previous discussion on this: https://github.com/rails/rails/pull/20440#discussion_r31867428

The bottom line is we're testing for nil values, which means we have to keep the backwards compatibility :smile:

sbhatore95 replied:
@kaspth I don't think you got my point. We are testing for nil values of signed messages, not the value to be signed itself. The above discussion is about passing `nil` to `valid_message?` message and not `generate` function, the one I am talking about :smile: 
   As far as backwards compatibility is concerned, I think it means that we should respect all the signed messages from previous(or rather the one in use) `MessageVerifier` which we will do. I am just saying that we might want to consider accepting `nil` as a value argument for `generate` function. It doesn't make sense to sign a `nil` value, which our `MessageVerifier` does. I hope this makes sense :smile:

kaspth replied:
Ah, I see. And I do agree with you that it doesn't make sense to sign a nil value when using the verifier directly. Could it be because of some of the cookie stuff?

Anyway if we want to change this we probably need to go through a deprecation cycle. Which might be a bit of a hassle instead of just keeping it around. Look at it from another point of view, if it doesn't make sense to sign a nil value, then people probably wouldn't do it. Then our deprecation cycle is for naught :smile:

sbhatore95 replied:
Ok, then let's go with the way it is :+1:

sbhatore95 commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,156 @@ def valid_message?(signed_message)
+    #
+    # JWS Signature  Signature over the JWS Header and the JWS Payload(Claims set).
+    # The signing input is given by ASCII(BASE64URL(UTF8(JWS Protected Header)) || '.' ||
+    # BASE64URL(JWS Payload)).
```
I have not done strictly UTF8 of the JWS protected header as written above. As far as I know, serialized(e.g Marshal.dump) data is also a  UTF8 string. Here http://tools.ietf.org/html/rfc7515#section-3.1

kaspth replied:
Well, users can use a different serializer, like JSON, but yes, I think it's safe to assume it's UTF-8. Why not add a test to be safe? You can use `encoding` on a string.

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,156 @@ def valid_message?(signed_message)
+    #
+    #   # After 1 hour...
+    #   verifier.verify(expiring_message, for: 'remember_me')
+    #   # => ActiveSupport::Claims::ExpiredClaims
```
We can't use `=>` as that implies output, we should say `raises ActiveSupport...`

sbhatore95 replied:
Yeah !! I got misled as the previous documentation did it the '=>' way. I will correct it.

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,156 @@ def valid_message?(signed_message)
+    #   verifier.verify(expiring_message, for: 'remember_me')
+    #   # => ActiveSupport::Claims::ExpiredClaims
+    #
+    # You can also use +expires_at+ instead of +expires_in+. Just replace +expires_in+ by +expires_at+
```
How about: `A message can also be set to expire at an explicit time with +expires_at+:` (and then an example)

Also you should read our documentation guidelines, we try to avoid the use of `you` and `your` for instance.

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # Raises +ExpiredClaims+ if the message is expired.
+    #
+    #   expired_message = verifier.generate('a private message', expires_at: 1.day.ago)
+    #   verifier.verified(signed_message) # raises ActiveSupport::Claims::ExpiredClaims
```
The idea behind `verified` is that it shouldn't raise anything, but return `nil`. We need to match the verifiers behavior in Claims:
- `verify` returns nil if the claims are invalid or expired.
- `verify!` raises

sbhatore95 replied:
Oh shoot !! You are right. The whole purpose of `verified`was that it shouldn't raise anything. I forgot about the raise by `Claims`.

kaspth replied:
No worries. That’s why we have reviews :smile:

> Den 10/07/2015 kl. 21.11 skrev Siddharth Bhatore notifications@github.com:
> 
> In activesupport/lib/active_support/message_verifier.rb https://github.com/sbhatore/rails/pull/3#discussion_r34386600:
> 
> > -        rescue ArgumentError => argument_error
> > -          return if argument_error.message =~ %r{invalid base64}
> > -          raise
> > -        end
> > -    #   incompatible_message = "aGVhZGVy.Y2xhaW1z.OWYzY2Y1Mzc1MmUxNjJlMzFhZDY3ODVlNGMyMTQ5YWE2OTc0NjQ5MQ=="
> > -    #   verifier.verified(incompatible_message) # raises TypeError: incompatible marshal file format
> > -    #
> > -    # Raises +InvalidClaims+ if the message doesn't have the same purpose as that
> > -    # given in the options.
> > -    #
> > -    #   verifier.verified(signed_message, for: 'something_else') # raises ActiveSupport::Claims::InvalidClaims
> > -    #
> > -    # Raises +ExpiredClaims+ if the message is expired.
> > -    #
> > -    #   expired_message = verifier.generate('a private message', expires_at: 1.day.ago)
> > -    #   verifier.verified(signed_message) # raises ActiveSupport::Claims::ExpiredClaims
> >   Oh shoot !! You are right. The whole purpose of verifiedwas that it shouldn't raise anything. I forgot about the raise by Claims.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/sbhatore/rails/pull/3/files#r34386600.

## 

Kasper

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
     end
 
-    # Generates a signed message for the provided value.
+    # Generates a signed message for the given +value+ and +options+ hash.
```
rm `hash` here.

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
     # The message is signed with the +MessageVerifier+'s secret. Without knowing
-    # the secret, the original value cannot be extracted from the message.
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed in the form of JSON Web Token(JWT). The final message is
```
`signed as a...`

Also put a space before the abbreviation parentheses (here and all other places): `JSON Web Token (JWT)`.

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed in the form of JSON Web Token(JWT). The final message is
+    # concatenation of base64-encoded JWS header, base64-encoded JWT claims set
+    # and  base64-encoded signature seperated by a period('.'). This format is
```
A space too much after `and`

kaspth replied:
`separated`

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # message is signed in the form of JSON Web Token(JWT). The final message is
+    # concatenation of base64-encoded JWS header, base64-encoded JWT claims set
+    # and  base64-encoded signature seperated by a period('.'). This format is
+    # known as JSON Web Signature(JWS) structure:
```
rm `structure` here.

kaspth replied:
`known as a`

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #
+    # ==== Options
+    #
+    # * <tt>:expires_at</tt> - Expiry time of the signed message.
```
`Explicit expiring time for the signed message.`

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # ==== Options
+    #
+    # * <tt>:expires_at</tt> - Expiry time of the signed message.
+    # * <tt>:expires_in</tt> - Time from now after which the message
```
`Relative time the message should expire after (e.g. 1.hour).`

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # * <tt>:expires_at</tt> - Expiry time of the signed message.
+    # * <tt>:expires_in</tt> - Time from now after which the message
+    #   will expire (e.g. 1.month).
+    # * <tt>:for</tt> - Purpose/Scope of the signed message (defaults to 'universal').
```
`The signed message purpose confines usage to places with the same purpose. Defaults to a universally usable message.`

We don't have to tell them about the `'universal'` value that we assign.

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #   verifier.generate('a private message')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIOdW5pdmVyc2FsBjsGVDoIZXhwSSIdMjAxNS0wOC0wOVQxODowNjowMy44NzhaBjsGVA==.ZmRlYjhlMTkxNGI0NGQ4OWRkNDdjZWM5YTY1NmQ3MTM5OWY5NzRmMQ=="
+    #
+    # Messages can be given a purpose/scope to bump the security up some more.
```
rm `/scope`

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIOdW5pdmVyc2FsBjsGVDoIZXhwSSIdMjAxNS0wOC0wOVQxODowNjowMy44NzhaBjsGVA==.ZmRlYjhlMTkxNGI0NGQ4OWRkNDdjZWM5YTY1NmQ3MTM5OWY5NzRmMQ=="
+    #
+    # Messages can be given a purpose/scope to bump the security up some more.
+    # In this way evildoers can't reuse a sign-up form's cookie on the login
```
`# This way evildoers can't reuse a message generated for the sign-up form on any other page:`

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #
+    #   signup_message = verifier.generate('a private message', for: 'signup_form')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIQc2lnbnVwX2Zvcm0GOwZUOghleHBJIh0yMDE1LTA4LTA5VDE4OjA1OjEwLjM2N1oGOwZU.YTMxMzIzMDc3YTI3ZDkwM2FmNWQ1MDgwNzAxNTgwMjNjNzI2NDE0Zg=="
+    #
```
Add a failing example:

``` ruby
#
#   verifier.verified(signup_message, for: 'something_else`) # => nil
#
```

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #   verifier.verify(signup_message, for: 'signup_form')
+    #   # => 'a private message'
+    #
+    # A message can be set that expires some time in future with +expires_in+.
```
``` ruby
# A message can also be given an expiration time with either +expires_in+ or +expires_at+.
# Useful when a message should not be accessible indefinitely:
```

kaspth replied:
Wait, let's redo as:

``` ruby
# A message can also be given an expiration time with either +expires_in+ or +expires_at+.
# Useful when a message should not be accessible indefinitely. The +expires_in+ option accepts a relative time,
# which is resolved to an explicit expiry time, when the message is signed:
```

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #
+    # A message can also be set to expire at an explicit time with +expires_at+:
+    #
+    #   expiring_message = verifier.generate('a private message', for: 'something', expires_at: 2.hours.from_now)
```
Using `from_now` makes this feel like the `expires_in` example. Let's use `Date.tomorrow.midnight`. Then...

kaspth commented on activesupport/lib/active_support/message_verifier.rb:
```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    #   verifier.verify(expiring_message, for: 'something')
+    #   # => "a private message"
+    #
+    #   # After 2 hours...
```
...this should be something like `# After the day is over`

sbhatore95 replied:
But the message expires in 2 hours, doesn't it ?

kaspth replied:
Sorry, I was trying to tie it with this message: https://github.com/sbhatore/rails/pull/3/files#r34387955 - but that only makes sense when you're seeing it in the web interface :smile:

sbhatore95 replied:
Oh,  I see :stuck_out_tongue:

kaspth commented on activesupport/lib/active_support/railtie.rb:
```diff
@@ -47,5 +47,11 @@ class Railtie < Rails::Railtie # :nodoc:
+
+    initializer "active_support.set_claims_expiration" do |app|
+      require 'active_support/claims'
+      app.config.active_support.claims_expiration ||= 1.month
```
I'm still not sure if we should expose Claims to general users (that's another question we should ask in the PR). Anyway, let's rename this to `messages_expires_in`.

sbhatore95 replied:
Well I think that `Claims` shouldn't be available to general public. It is just like a helper to `MessageVerifier`. `Claims` is not much of use by itself.

kaspth commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,57 @@
+require 'active_support/time'
+
+module ActiveSupport
+  class Claims
```
Given we want Claims to be private - https://github.com/sbhatore/rails/pull/3/files#r34388981 - we need to add a `:nodoc:` so it won't show up in the generated documentation.

Like this:

``` ruby
class Claims # :nodoc:
```

sbhatore95 replied:
Sure ! I was thinking of bringing that up. Thanks :+1:

sbhatore95 replied:
Also, we would not want people to have a look at `LegacyMessageVerifier` either, right ? So, a  `:nodoc:` for that too would be needed.

kaspth replied:
Yes, but that’s defined under `private` right? Because I believe stuff under there doesn’t show up in the documentation. But I don’t know if that’s true for classes. Eh, let’s add one to be safe. Then people can always see it’s not needed :smile:

> Den 10/07/2015 kl. 22.02 skrev Siddharth Bhatore notifications@github.com:
> 
> In activesupport/lib/active_support/claims.rb https://github.com/sbhatore/rails/pull/3#discussion_r34390948:
> 
> > @@ -0,0 +1,57 @@
> > +require 'active_support/time'
> > +
> > +module ActiveSupport
> > -  class Claims
> >   Also, we would not want people to have a look at LegacyMessageVerifier either, right ? So, a :nodoc: for that too would be needed.
> 
> —
> Reply to this email directly or view it on GitHub https://github.com/sbhatore/rails/pull/3/files#r34390948.

## 

Kasper

sbhatore95 replied:
It's actually showing up in the documentations :smile: I will add a `:nodoc:` there then.

kaspth commented on activesupport/lib/active_support/claims.rb:
```diff
@@ -0,0 +1,69 @@
+    class << self
+      attr_accessor :expires_in
+
+      def verify!(claims, options = {})
```
I've been thinking about this some more. How about doing it this way:

``` ruby
def verify!(claims, options = {})
  claims[:pld] if validate(claims, options, raises: true)
end

def verify(claims, options = {})
  claims[:pld] if validate(claims, options)
end

private
  def validate(claims, options, raises: false)
    if raises
      same_purpose?(claims[:for], options[:for]) || raise InvalidClaims
      fresh?(claims[:exp]) || raise ExpiredClaims
      true
    else
      equal_purpose?(claims[:for], options[:for]) && fresh?(claims[:exp])
    end
  end

  def same_purpose?(claims_purpose, other_purpose)
    claims_purpose == other_purpose || UNIVERSAL_PURPOSE
  end

  def fresh?(expiration)
    Time.iso8601(expiration) < Time.now.utc if expiration
  end
```

Then all the really nasty bits are in `validate`, which can be reused in `verify!` and `verify`. Later if we want to add extra metadata only `validate` has to change.

I'd certainly love it if `validate` could be more readable :smile:

What do you think?

kaspth replied:
I'm also not in love with the `same_purpose?` name, so if you've got something better please tell me :smile:

kaspth replied:
Ah crap, actually I'm sorry. I've given you an answer instead of poking you with questions, so you could find a solution. Feel free to disregard what I've said above and come up with your own solution :smile:

My main idea behind reworking this was that the exception raising (or not) spread all over the place, and it would be tough to add extra metadata later. So finding a place to put all that junk would make the code easier to change later :smile:

sbhatore95 replied:
I thought upon it for quite a while @kaspth. I still don't understand why we want to convert what we are doing in `verify` and `verify!` into `if` and `else` of `validate` :smile: . We would be able to add extra metadata afterwards, and check for them in `verify` and `verify!` functions. Whatever we do, we have to repeat some code. Be it in `if` and `else` of `validate` or directly in `verify` and `verify!` methods.

Anyways, I have made things more readable like this. What do you think about this? :

``` ruby
      def verify!(claims, options = {})
        raise_if_invalid_purpose(claims[:for], options[:for])
        claims[:pld] if raise_if_expired(claims[:exp])
      end

      def verify(claims, options = {})
        claims[:pld] if check_purpose?(claims[:for], options[:for]) && not_expired?(claims[:exp])
      end

      private
        def not_expired?(expiration)
          return true unless expiration

          Time.iso8601(expiration) > Time.now.utc
        end

        def check_purpose?(claims_purpose, other_purpose)
          claims_purpose == (other_purpose || UNIVERSAL_PURPOSE)
        end

        def raise_if_invalid_purpose(claims_purpose, other_purpose)
          check_purpose?(claims_purpose, other_purpose) || raise(InvalidClaims)
        end

        def raise_if_expired(expiration)
          not_expired?(expiration) || raise(ExpiredClaims)
        end
```

sbhatore95 replied:
I think `check_purpose?` can be used instead of `same_purpose?`.

kaspth replied:
Sure, we can still get at them in the verify methods, but the nice thing about having that stuff contained in `validate` is that anything we add there will automatically improve the verify methods. This means the verify methods don't have to **change**.

To me your changes add too much indirection by putting everything into its own method. What I liked about `validate` was that my eye would scan `verify` and `verify!` (which was easy because they're one line) and see I had to look in `validate`. Looking in `validate` I could see everything I needed at a glance because the helper methods were clearly named - I could even stop reading there.

Without having that `validate` method I find there's too many pieces and the extra method "boundaries" makes it tougher to keep in my head.

`check_purpose?` doesn't capture it for me. What does it mean for a purpose to be checked? How does it pass and how does it fail? `same_purpose?` is clearer IMO.

I also think `unexpired?` is better than `not_expired?`. Though I think it would be better to find a word that focuses on the freshness - you know instead of saying unexpired what's the "positive" way of saying the token is still alive.

sbhatore95 replied:
> > Sure, we can still get at them in the verify methods, but the nice thing about having that stuff contained in validate is that anything we add there will automatically improve the verify methods. This means the verify methods don't have to change.

Ok, I am also inclined to using `validate` now. Not having to change `verify` and `verify!` would be better. :+1: 

> > check_purpose? doesn't capture it for me. What does it mean for a purpose to be checked? How does it pass and how does it fail? same_purpose? is clearer IMO.

By `check_purpose`, I meant that in the `verify` method of `Claims`, we are basically checking if the purpose we have in ``claims` matches with the one in the `options`(entered in `verify` of `MessageVerifier`). If the purposes don't match, then we are raising an exception. I don't think there is any harm in going with either of these :smile:.  

> > I also think unexpired? is better than not_expired?. Though I think it would be better to find a word that focuses on the freshness - you know instead of saying unexpired what's the "positive" way of saying the token is still alive.

That's an interesting thought.

kaspth commented:
Ok, I've added one comment. Send a PR after you've taken a look at that :smile:

sbhatore commented:
Thanks @kaspth . I agree with you about the above comment. It would be better to have the validation under one method to support extendability. I will take your input into consideration, think upon it some more and open a PR today itself.
   About the commit message of my PR, how does this look
https://github.com/sbhatore/rails/commit/ad99ccc0cdbdb2177f6b5cdce608feb5ff9cb86c

sbhatore95 committed:
MessageVerifer:Signing of message in JWS format

Any message can be signed now in JSON Web Signature (JWS) with JSON Web Token (JWT) Claims set
as its payload. The JWT Claims set contains 'pld' - the message, 'exp' - expiration and
'for' - purpose of the message. Now, a message can be expired on time internally and it can
also be given a purpose so that evildoers are not able to misuse a message in place of other.
JWT Claims is not fully implemented yet as per the spec here:
http://tools.ietf.org/html/draft-ietf-oauth-json-web-token-13#section-4
This is just a basic implementation of JWT and JWS which could be extended in future.

sbhatore commented:
I am merging this and opening a PR in rails :smile:
